test/other/embind_tsgen_es6.d.ts

```diff
@@ -0,0 +1,114 @@
+  functionFive(x: number, y: number): number;
+  constFn(): number;
+  longFn(_0: number): number;
+  functionThree(_0: ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string): number;
+  functionSix(str: ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string): number;
+  delete(): void;
+}
+
+export interface BarValue<T extends number> {
+  value: T;
+}
+export type Bar = BarValue<0>|BarValue<1>|BarValue<2>;
+
+export interface EmptyEnumValue<T extends number> {
+  value: T;
+}
+export type EmptyEnum = never/* Empty Enumerator */;
+
+export type ValArrIx = [ Bar, Bar, Bar, Bar ];
+
+export interface IntVec {
+  push_back(_0: number): void;
+  resize(_0: number, _1: number): void;
+  size(): number;
+  get(_0: number): number | undefined;
+  set(_0: number, _1: number): boolean;
+  delete(): void;
+}
+
+export interface MapIntInt {
+  keys(): IntVec;
+  get(_0: number): number | undefined;
+  set(_0: number, _1: number): void;
+  size(): number;
+  delete(): void;
+}
+
+export interface Foo {
+  process(_0: Test): void;
+  delete(): void;
+}
+
+export type ValObj = {
+  foo: Foo,
+  bar: Bar
+};
+
+export interface ClassWithConstructor {
+  fn(_0: number): number;
+  delete(): void;
+}
+
+export interface ClassWithTwoConstructors {
+  delete(): void;
+}
+
+export interface ClassWithSmartPtrConstructor {
+  fn(_0: number): number;
+  delete(): void;
+}
+
+export interface BaseClass {
+  fn(_0: number): number;
+  delete(): void;
+}
+
+export interface DerivedClass extends BaseClass {
+  fn2(_0: number): number;
+  delete(): void;
+}
+
+export type ValArr = [ number, number, number ];
+
+interface EmbindModule {
+  Test: {staticFunction(_0: number): number; staticFunctionWithParam(x: number): number; staticProperty: number};
+  class_returning_fn(): Test;
+  class_unique_ptr_returning_fn(): Test;
+  a_class_instance: Test;
+  an_enum: Bar;
+  Bar: {valueOne: BarValue<0>, valueTwo: BarValue<1>, valueThree: BarValue<2>};
+  EmptyEnum: {};
+  enum_returning_fn(): Bar;
+  IntVec: {new(): IntVec};
+  MapIntInt: {new(): MapIntInt};
+  Foo: {};
+  ClassWithConstructor: {new(_0: number, _1: ValArr): ClassWithConstructor};
+  ClassWithTwoConstructors: {new(): ClassWithTwoConstructors; new(_0: number): ClassWithTwoConstructors};
+  ClassWithSmartPtrConstructor: {new(_0: number, _1: ValArr): ClassWithSmartPtrConstructor};
+  BaseClass: {};
+  DerivedClass: {};
+  a_bool: boolean;
+  an_int: number;
+  optional_test(_0: Foo | undefined): number | undefined;
+  global_fn(_0: number, _1: number): number;
+  smart_ptr_function(_0: ClassWithSmartPtrConstructor): number;
+  smart_ptr_function_with_params(foo: ClassWithSmartPtrConstructor): number;
+  function_with_callback_param(_0: (message: string) => void): number;
+  string_test(_0: ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string): string;
+  wstring_test(_0: string): string;
+}
+export type MainModule = WasmModule & EmbindModule;
+export type MainModuleFunctor = () => Promise<MainModule>;
```

Comment:
Ideally this would have all the options as well, (e.g. `locateFile`, not sure if there are others off-hand), but that kinda begs your original bigger issue of https://github.com/emscripten-core/emscripten/issues/20296 ... I'm unopinionated if this is more or less helpful without the args.

Reply:
Agreed on that. My hope is this small change will be enough to allow users to generate a node `package.json` around wasm/js/d.ts output of `"type": "module", so that some wrapper library consuming the wasm modules can easily make use of the TypeScript declarations.

I can add `locateFile`, but am not sure of a good way to test that. Any thoughts there?

Reply:
well, it'd be easy enough to add locateFile as a one-off optional option here, but it seems like there's a whole pile of them (are they these? https://github.com/emscripten-core/emscripten/blob/e8262f62974a3de2755a210869b8aeb6279d2c07/src/settings.js#L951 )

Maybe the "quick win" that you're looking for is to just change this to `export type MainModuleFunctor = (options: unknown) => Promise<MainModule>;` so that it indicates options exist without having to tackle the larger problem of enumerating/typing them all.

I'm also not super knowledgeable about all this code, just a fellow interested onlooker, so I appreciate you taking a stab at it!

Reply:
That seems like a great start to me. Will get that added shortly.

And thanks for the feedback! 🙇